src/PerimeterxFirstPartyClient.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace Perimeterx;
+
+final class PerimeterxFirstPartyClient {
+    const CLIENT_DOMAIN = 'client.perimeterx.net';
+    const CAPTCHA_DOMAIN = 'captcha.px-cdn.net';
+    const XHR_DOMAIN = 'collector.perimeterx.net';
+    const FIRST_PARTY_HEADER_NAME = 'X-PX-First-Party';
+    const FIRST_PARTY_HEADER_VALUE = '1';
+
+    const EMPTY_GIF_B64 = 'R0lGODlhAQABAPAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';
+
+    private $pxConfig;
+    private $logger;
+    private $httpClient;
+
+    /**
+     * @param array $pxConfig
+     * @param Perimeterx\Utils\IHttpClient $httpClient
+     */
+
+    public function __construct($pxConfig, $httpClient) {
+        $this->pxConfig = $pxConfig;
+        $this->logger = $pxConfig['logger'];
+        $this->httpClient = $httpClient;
+    }
+
+    public function handleFirstParty() {
+        $appIdSubstr = substr($this->pxConfig['app_id'], 2);
+        $initJsPath = "/$appIdSubstr/init.js";
+        $xhrPath = "/$appIdSubstr/xhr";
+        $captchaJsPath = "/$appIdSubstr/captcha";
+
+        $uri = strtok($_SERVER['REQUEST_URI'], "?");
+        $method = $_SERVER['REQUEST_METHOD'];
+
+        $retval = null;
+
+        if ($uri === $initJsPath) {
+            $retval = $this->handleFirstPartyClient();
+        } else if (strpos($uri, $xhrPath) !== false) {
+            $retval = $this->handleFirstPartyXHR($xhrPath, $method, $uri);
+        } else if (strpos($uri, $captchaJsPath) !== false) {
+            $retval = $this->handleFirstPartyCaptcha();
+        } else {
+            $this->logger->debug("$method $uri did not match any first party path");
+        }
+        return $retval;
+    }
+
+    private function handleFirstPartyClient() {
+        $this->logger->debug('first party client script detected');
+        return $this->handleFirstPartyScript(self::CLIENT_DOMAIN, "/{$this->pxConfig['app_id']}/main.min.js");
+    }
+
+    private function handleFirstPartyCaptcha() {
+        $this->logger->debug('first party captcha script detected');
+        return $this->handleFirstPartyScript(self::CAPTCHA_DOMAIN, "/{$this->pxConfig['app_id']}/captcha.js");
+    }
+
+    private function handleFirstPartyXHR($xhrPath, $method, $uri) {
+        $this->logger->debug('first party xhr detected');
+
+        $thirdPartyUri = str_replace($xhrPath, '', $uri);
+        $isGifRequest = strpos($thirdPartyUri, ".gif") !== false;
```

Comment:
is this logic happens before filter by extension?  who should Overcome who?

Reply:
This happens before filtering -- that's how it's implemented in node-core, which was my basis.

Reply:
Good thats what it should be 🥇